### ISSUE TITLE
time to handle hooks

### ISSUE THREAD

**Issue by StoneCypher** [StoneCypher/fsl#255]

**Comment by StoneCypher** [comment]
@machinshin

**Comment by StoneCypher** [comment]
Remember, we lost a user over this

Get this done

https://github.com/StoneCypher/jssm/issues/389

**Comment by theshadow27** [comment]
I was also very excited about the fssm project - especially the DSL and graphing - until I realized there were no callback functions/eventEmitter support. Graphical view for system validation is extremely nice, but without events there is so much glue that it looses it's advantage over something like [Finity](https://npm.io/package/finity) with no DSL and difficult to read source but native event support. 

I think the hardest part is going to be incorporating into the DSL... though looking through samples, I think that [the extreme test machine](https://github.com/StoneCypher/jssm/blob/main/src/machines/extreme%20case.fsl) the `${functionName}` syntax is pretty elegant. 

An alternate approach would be to forgo DSL entirely and just populate the Machine object with functions per the state name. This has the advantage of boilerplate reduction, convention over configuration, and keeping the DSL tight. For example, the [Javascript-State-Machine](https://www.npmjs.com/package/javascript-state-machine) project just invokes functions if present, and adds accessors (this can probably be done with a proxy, or by actually adding them during parse). 

> .. methods to transition to a different state:
> fsm.melt()
> fsm.freeze()
> fsm.vaporize()
> fsm.condense()
> ... observer methods called automatically during the lifecycle of a transition:
> onMelt()
> onFreeze()
> onVaporize()
> onCondense()

presumably you'd also want `beforeMelt()`, `onMelt()`, `afterMelt()` .. .

If Machine inherits EventEmitter (which is pretty cheap) then the default 'onState' functions can invoke `this.emit('State')` etc. 

If all of this seems too expensive for cases where it is unused, then a small DSL extension like 
```
events: on;

--- or ---

events: [freeze melt]
before-events: [condense]
```
would let users opt in... or a shorthand (RegExp-ish `/e` for instance) to "just turn it all on" to maintain the beautiful (mega props, BTW) one liners:

```
const TrafficLightsEventEmitter = sm`Red -> Green -> Yellow -> Red; /e`;
```

Just some thoughts. On a tight deadline so I will have to use something else, but I hope to use (~> contrib?) to fssm in the future. Great work thus far!!

**Comment by StoneCypher** [comment]
that's actually almost exactly the planned api (enter rather than emit because of ISO naming)

the reason i've been dragging my heels is that this is going to be cross-language, and defining that in a cross language way is brutal

i'm thinking about opting out tbh and mumbling some "the implementation of which is left to the reader" type nonsense

**Comment by theshadow27** [comment]
yeah I saw your cross-language tickets... very ambitious... but at least in node (jssm) it is a very tough sell w/o any support. 

The EventEmitter->jssm is low priority, one line of code could bind all actions... 

```
const fsm = sm`Red -> Green -> Yellow -> Red; `;
const bus = new EventEmitter(fsm);

fsm.list_actions().forEach( a => bus.on(a, bus.action.bind(bus, a) ) );
```

but a correct jssm->EventEmitter is not possible without completely re-implementing the `action`/`transition`/`force_transition` functions. 

would you be open to at least adding a couple empty functions on `Machine`, like

```
guard(what: string, action: StateType, lastState: StateType, nextState: StateType, data?: mDT): boolean {
   return true; // TODO implement events+guards
}
mutating(direction: string, state: StateType,  data?: mDT): void {}

```

and stubbing out 

```
  action(name: StateType, newData?: mDT): boolean {

    if (this.valid_action(name, newData)) {
      const edge: JssmTransition<mDT> = this.current_action_edge_for(name);

      if(!this.guard( 'action' , name, this._state, edge.to, newData ) )  return false; /* NEW! */

      this.mutating('exit', this._state, newData);
      this._state = edge.to;
      this.mutating('enter', this._state, newData);
      
      return true;
    } else {
      this.guard('invalid_action', name, this._state, null, newData);  /* NEW just a callback */
      return false;
    }
  }

  transition(newState: StateType, newData?: mDT): boolean {
    if (this.valid_transition(newState, newData)) {

      if(!this.guard( 'transition' , 'transition', this._state, edge.to, newData ) )  return false; /* NEW! */

      this.mutating('enter', this._state, newData);   /* NEW! */
      this._state = newState;
      this.mutating('exit', this._state, newData);     /* NEW! */

      return true;
    } else {
      this.guard( 'invalid_transition' , 'transition', this._state, newState, newData );  /* NEW! just a callback */
      return false;
    }
  }

// same for forced_transition

```

stays synchronous/deterministic, no changes to the DSL, only two new tests to keep 100% coverage (again, major props there), and the pattern is portable across any language with inheritance/prototyping. In ES, the JIT will immediately prune the branches and empty functions -- I'd bet $1 there would be no affect on performance. And then, for those readers so inclined, we could have decent event support with three nominally-one-liners:

```
const fsm = sm`Red -> Green -> Yellow -> Red; `;
const bus = new EventEmitter(bus);

fsm.list_actions().forEach( a => bus.on(a, bus.action.bind(bus, a) ) );
fsm.mutating=(direction, state,  data)=>{  bus.emit(direction, state, data );  if(direction==='enter') bus.emit(state, data); }
fsm.guard = (what, action, lastState, nextState, data) => { 
   try { 
       bus.emit( what.startsWith('invalid_') ? what :  `>${action}` , {action, data, lastState, nextState}) ;
       return true;
   } catch(e){return false;}
};
```

or something like that... anyway, just my $0.02. I'd offer a PR but I'm not a TypeScript guy and the code is too pretty for me to butcher. 

p.s. I ended up going with Finity not just for events but also because of the async support and builtin timeouts, which the above doesn't cover (though it is also easy to shim). Have to say after playing with jssm it is EXTREMELY annoying not being able to paste the code into a web-based viewer to verify the diagram. So quit dragging your feet already ;P but really, perfect is the enemy of good enough.

**Comment by StoneCypher** [comment]
Got asked again by another new user in email today

Really need to get this done

**Comment by theshadow27** [comment]
Eye of the beholder 😉

But really, the simplicity is very nice.

Looking forward to see what you come up with. No pressure hah

**Comment by StoneCypher** [comment]
@theshadow27 - hooks on fully named transitions and on actions have landed

```
const tl = sm`Red => Green => Yellow => Red;`
  .hook('Yellow', 'Red', () => console.log('Red light!');
```

**Comment by StoneCypher** [comment]
and for the record i ended up implementing almost everything you suggested, in ways quite similar to what you suggested

your recommendations were solid
